### PR TITLE
Validate Data Quality Rules

### PR DESCRIPTION
**Pull request by ManasaKanathur**
This PR introduces validations for each data quality rule type – row, query, aggregate to ensure that expectations are well-formed and semantically correct for that rule type before any further processing. Custom exceptions are raised for each validation failure.  

## Description
The validate_rules.py module defines the SparkExpectationsValidateRules class, which provides static methods to validate different types of data quality rules (row_dq, agg_dq, and query_dq) before they are executed on Spark DataFrames. It distinguishes between row_dq, agg_dq, and query_dq rule types, using sqlglot to parse and check expectations for correct format and usage (e.g., no aggregates in row_dq, required aggregates in agg_dq, valid SQL for query_dq). It returns a mapping of failed rules by type, enabling robust and maintainable DQ rule validation. 

## Motivation and Context
The exception is currently raised at a later stage in the processing pipeline, which means invalid expectations are only detected after Spark attempts to execute them. To improve robustness and fail-fast behavior, validation should occur earlier in the workflow—ideally before execution—so that malformed or inappropriate expectations are caught and reported immediately, reducing unnecessary computation and making error handling more predictable and user-friendly.

## How Has This Been Tested?
The test_validate_rules.py file contains unit tests for the validation logic in validate_rules.py. It uses pytest to check that the validation methods correctly accept valid data quality rules and raise the appropriate exceptions for invalid or malformed rules. The tests cover different rule types (row_dq, agg_dq, query_dq), using sample Spark DataFrames and various expectation strings to ensure robust rule validation and error handling.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: spark_expectations/utils/validate_rules.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+            df (DataFrame): The input DataFrame.
+            rule (Dict): Rule containing the 'expectation' and 'rule'.
+
+        Raises:
+            SparkExpectationsInvalidRowDQExpectationException: If aggregate functions are used or expression fails.
+        """
+        expectation = rule.get("expectation").lower()
+        disallowed_funcs = ["sum", "avg", "min", "max", "stddev", "count"]
+
+        if any(func in expectation for func in disallowed_funcs):
```
Comment (smirkovic) [inline spark_expectations/utils/validate_rules.py]: Are we expecting to see something like `customer_id is not null` as value for `expectation` and how will that work for this check ? 

`validate_rules.py` is perfect for adding unit tests so you can easily test your logic

Comment (ManasaKanathur) [reply spark_expectations/utils/validate_rules.py]: Will add unit tests for validation utility; will also include a unit test with expectation as "customer_id" is not null.

**Comment by liorz901** [comment]
I noticed that some of the commits are not verified. Will that cause an issue we are ready to merge this.

File: spark_expectations/utils/validate_rules.py [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+        temp_view_name = f"__dq_validation_temp_{uuid.uuid4().hex}__"
+        # Replace {placeholder} table names
+        query = re.sub(r"\{[^}]+\}", temp_view_name, query)
+
+        # Extract table names
+        table_names = SparkExpectationsValidateRules.extract_table_names_from_sql(query.lower())
+
+        # Replace all extracted table names in the query with the temp view name
+        for table_name in table_names:
+            query = re.sub(rf"\b{re.escape(table_name)}\b", temp_view_name, query)
```
Comment (smirkovic) [inline spark_expectations/utils/validate_rules.py]: What happens if we have nested queries ? If those individual queries have different table names we are replacing all table names with same `temp_view_name`

File: spark_expectations/utils/validate_rules.py [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+
+        try:
+            # Create a temporary view for validation
+            df.createOrReplaceTempView(temp_view_name)
+            # Validate the query
+            spark.sql(f"SELECT {query} AS result").collect()
```
Comment (smirkovic) [inline spark_expectations/utils/validate_rules.py]: is it possible to capture exception on existing `expectation` execution flow or we had to create new TempView for every rule ? I feel like it can end up that we put a lot of extra burden on DBX driver

**Comment by liorz901** [comment]
> > I noticed that some of the commits are not verified. Will that cause an issue we are ready to merge this.
> 
> Meaning some checks not being successful for some of the commits? No that should not cause an issue because I updated the code to ensure build was successful.

It has to do with GPG key/email used for those commits.

File: spark_expectations/utils/validate_rules.py (hunk 2 above)
Comment (ManasaKanathur) [reply spark_expectations/utils/validate_rules.py:128]: Yes if we have nested queries / queries with more than one table, we are replacing them with the same temp_view_name because we don't want to create multiple views.

Comment (ManasaKanathur) [reply spark_expectations/utils/validate_rules.py]: Yes if we have nested queries / queries with more than one table, we are replacing them with the same temp_view_name because we don't want to create multiple views (increase load in dbx driver)

Since the goal here is to just validate the query itself and ensure it executes on the given spark dataframe, replacing all referenced tables with same temp view works in this context.

File: spark_expectations/utils/validate_rules.py (hunk 3 above)
Comment (ManasaKanathur) [reply spark_expectations/utils/validate_rules.py]: Capturing exception on existing expectation execution flow defeats the purpose of this new feature where we are trying to handle exceptions at an earlier stage in processing. 

You make a very good point on DBX burden. I have an idea on how to go about this......can create a single shared temp view and reuse it for all query rules instead of one temp view per query rule. Another option is using regex to check if it is a valid query instead of actually running on dataframe (no creation of temp view)

File: spark_expectations/utils/validate_rules.py [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+            df (DataFrame): The input DataFrame.
+            rule (Dict): Rule containing the 'expectation' and 'rule'.
+
+        Raises:
+            SparkExpectationsInvalidRowDQExpectationException: If aggregate functions are used or expression fails.
+        """
+        expectation = rule.get("expectation").lower()
+        disallowed_funcs = ["sum", "avg", "min", "max", "stddev", "count"]
+
+        if any(func in expectation for func in disallowed_funcs):
```
[the remark this answers is not in the document]
Comment (ManasaKanathur) [reply spark_expectations/utils/validate_rules.py]: does not work

File: spark_expectations/utils/validate_rules.py (hunk 3 above)
Comment (smirkovic) [reply spark_expectations/utils/validate_rules.py]: @ManasaKanathur 🤔 maybe we can use `EXPLAIN query`  to check validity without actually running expected queries (https://spark.apache.org/docs/latest/sql-ref-syntax-qry-explain.html) 

@asingamaneni 👀 what is your take on this PR

Comment (ManasaKanathur) [reply spark_expectations/utils/validate_rules.py]: Problem with `EXPLAIN` query on expressions like col1 > 20 or sum(col1) < 25, will not necessarily fail, because Spark SQL can generate a logical plan for these expressions—even though they are not valid or meaningful as standalone query_dq rules (since query_dq is expected to be a full SQL query e.g. <SELECT ... FROM ...> and not just an expression). It may be valid in the context of agg_dq or row_dq, but not for query_dq.

However, we can use explain() as a syntax and plan check, not as a full validator for rule type. And can include additional validation logic using regex to ensure that a query_dq expectation is a proper SQL query checking presence of both SELECT and FROM keywords in the expectation string before running explain().

Comment (ManasaKanathur) [reply spark_expectations/utils/validate_rules.py]: @asingamaneni what are your thoughts on these approaches?
1. i've implemented one already i.e creation of temp views; one per query rule (too much load on dbx driver)
2. alternatively can create a single shared temp view used by all rules (reduces burden on dbx driver)
3. use explain query along with regex to validate query_dq rule (no creation of temp views as we won't be running query on dataframe like the above 1. and 2.)

File: pyproject.toml [hunk 5]
```diff
@@ -15,7 +15,11 @@ classifiers = [
 ]
 # https://www.python.org/doc/versions/
 requires-python = ">=3.9,<=3.13"
-
+dependencies = [
+  "pluggy>=1",
+  "requests>=2.28.1",
+  "sqlglot",
```
Comment (smirkovic) [inline pyproject.toml]: let's move sqlglot where other dependencies are defined in pyproject.toml and also let's use some version constraints for the library

File: spark_expectations/utils/validate_rules.py [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+
+        Raises:
+            SparkExpectationsInvalidRowDQExpectationException: If aggregate functions are used or expression fails.
+        """
+        expectation = rule.get("expectation", "").lower()
+
+        if any(func in expectation for func in AGGREGATE_FUNCTIONS):
+            raise SparkExpectationsInvalidRowDQExpectationException(
+                f"[row_dq] Rule '{rule.get('rule')}' contains aggregate function (not allowed in row_dq): {expectation}"
+            )
```
Comment (smirkovic) [inline spark_expectations/utils/validate_rules.py]: Since we are leveraging `sqlglot` library we can use it's parser to verify if expression is using aggregate functions. 

That way we:
- don't need to do parsing ourself
- no need to store enum with possible aggregate function

```
expression = "avg(col2) > 18 and avg(col2) >20"

distinct_agg_functions = list({node.key for node in parse_one(expression).find_all(expressions.AggFunc)})
```


At the moment I believe sqlglot will return as AGG if any of the following functions are used

```
['bitwiseandagg', 'bitwiseoragg', 'bitwisexoragg', 'bitwisecountagg', 'parameterizedagg', 'argmax', 
'argmin', 'approxtopk', 'anonymousaggfunc', 'hll', 'approxdistinct', 'arrayagg', 'arrayuniqueagg', 
'arrayconcatagg', 'arrayunionagg', 'avg', 'anyvalue', 'lag', 'lead', 'first', 'last', 'firstvalue', 'lastvalue', 'nthvalue', 'count', 'countif', 'groupconcat', 'jsonobjectagg', 'jsonbobjectagg', 'logicalor', 'logicaland', 'max', 
'median', 'min', 'percentilecont', 'percentiledisc', 'quantile', 'sum', 'stddev', 'stddevpop', 'stddevsamp', 
'corr', 'variance', 'variancepop', 'covarsamp', 'covarpop']

```

**Review by yankomo — APPROVED** [review]

**Review by kyakkati — DISMISSED** [review]

File: pyproject.toml (hunk 5 above)
Comment (smirkovic) [reply pyproject.toml]: @ManasaKanathur one last minor thing you forgot to move sqlglot appropriate spot  

https://github.com/Nike-Inc/spark-expectations/blob/834db461da130965eb62d22ea86e0161fba2405e/pyproject.toml#L81
